File: sdk/samples/04_discovery/python/simple_discovery.py
#!/usr/bin/env python3
# SPDX-License-Identifier: Apache-2.0 OR MIT
# Copyright (c) 2025-2026 naskel.com

"""
Simple Discovery Sample - Demonstrates automatic multicast discovery

This sample shows how DDS participants automatically discover each other
using SPDP (Simple Participant Discovery Protocol) over multicast.

Run multiple instances to see them discover each other:
  Terminal 1: python3 simple_discovery.py
  Terminal 2: python3 simple_discovery.py

Key concepts:
- Automatic peer discovery via multicast
- No manual configuration required
- Domain ID for logical separation
"""

import os
import sys
import time

# Add SDK to path
sys.path.insert(0, '../../../python')

import hdds


def main():
    print("=== HDDS Simple Discovery Sample ===\n")

    # Get instance ID from args or generate random
    if len(sys.argv) > 1:
        instance_id = int(sys.argv[1])
    else:
        instance_id = os.getpid()

    print(f"Instance ID: {instance_id}")
    print("Domain ID: 0 (default)\n")

    # Initialize logging
    hdds.logging.init(hdds.LogLevel.INFO)

    # Create participant with default discovery settings
    print("Creating DomainParticipant...")
    participant = hdds.Participant(f"SimpleDiscovery_{instance_id}")

    print(f"[OK] Participant created: {participant.name}")
    print(f"     Domain ID: {participant.domain_id}")

    # Create writer and reader for discovery demo topic
    print("Creating DataWriter...")
    writer = participant.create_writer("DiscoveryDemo")
    print("[OK] DataWriter created")

    print("Creating DataReader...")
    reader = participant.create_reader("DiscoveryDemo")
    print("[OK] DataReader created")

    # Create waitset for efficient waiting
    waitset = hdds.WaitSet()
    waitset.attach_reader(reader)

    print("\n--- Discovery in Progress ---")
    print("Waiting for other participants to join...")
    print("(Run another instance of this sample to see discovery)\n")

    # Announce ourselves periodically
    announce_interval = 2.0  # seconds
    announce_count = 0

    try:
        while announce_count < 10:
            # Send an announcement
            announce_count += 1
            message = f"Hello from instance {instance_id} (message #{announce_count})"

            try:
                writer.write(message.encode('utf-8'))
                print(f"[SENT] {message}")
            except Exception as e:
                print(f"[ERROR] Failed to send: {e}")

            # Check for messages from other participants using waitset
            # Non-blocking poll (timeout=0)
            if waitset.wait(timeout=0.1):
                # Take all available samples
                while True:
                    data = reader.take()
                    if data is None:
                        break
                    print(f"[RECV] {data.decode('utf-8')}")

            # Wait before next announcement
            time.sleep(announce_interval)

    except KeyboardInterrupt:
        print("\n--- Interrupted ---")

    print("\n--- Sample complete (10 announcements sent) ---")
    print("\n=== Sample Complete ===")

    # Cleanup
    waitset.close()
    participant.close()

    return 0


if __name__ == "__main__":
    sys.exit(main())

File: sdk/samples/04_discovery/rust/src/bin/simple_discovery.rs
// SPDX-License-Identifier: Apache-2.0 OR MIT
// Copyright (c) 2025-2026 naskel.com

//! # HDDS Sample: Simple Discovery (SPDP)
//!
//! Demonstrates **automatic multicast discovery** using SPDP (Simple Participant
//! Discovery Protocol) - the zero-configuration way to find peers.
//!
//! ## How SPDP Works
//!
//! ```text
//! ┌─────────────────────────────────────────────────────────────────────┐
//! │                        Multicast Group                              │
//! │                      (239.255.0.1:7400)                             │
//! │                                                                     │
//! │   ┌──────────────┐         ┌──────────────┐        ┌─────────────┐ │
//! │   │ Participant A│◄───────►│ Participant B│◄──────►│Participant C│ │
//! │   │              │         │              │        │             │ │
//! │   │ GUID: 0xA... │         │ GUID: 0xB... │        │ GUID: 0xC...│ │
//! │   └──────────────┘         └──────────────┘        └─────────────┘ │
//! │                                                                     │
//! │   1. Each participant announces itself periodically                 │
//! │   2. Announcements include GUID, locators, QoS                      │
//! │   3. Peers receive announcements and update their database          │
//! └─────────────────────────────────────────────────────────────────────┘
//! ```
//!
//! ## Discovery Phases
//!
//! | Phase | Protocol | Purpose                          |
//! |-------|----------|----------------------------------|
//! | SPDP  | Multicast| Find participants in domain      |
//! | SEDP  | Unicast  | Exchange endpoint information    |
//!
//! ## Domain IDs
//!
//! ```text
//! Domain 0:              Domain 1:              Domain 2:
//! ┌──────────────┐      ┌──────────────┐       ┌──────────────┐
//! │ App A        │      │ App C        │       │ App E        │
//! │ App B        │      │ App D        │       │              │
//! └──────────────┘      └──────────────┘       └──────────────┘
//!     Isolated              Isolated              Isolated
//!
//! Different domains use different multicast ports - no cross-talk!
//! ```
//!
//! ## Running the Sample
//!
//! ```bash
//! # Terminal 1 - First participant
//! cargo run --bin simple_discovery
//!
//! # Terminal 2 - Second participant (will discover first)
//! cargo run --bin simple_discovery
//!
//! # Watch them discover each other automatically!
//! ```

use std::thread;
use std::time::Duration;

fn main() -> Result<(), Box<dyn std::error::Error>> {
    println!("=== HDDS Simple Discovery Sample ===\n");

    // Get instance ID from args or generate random
    let instance_id: u32 = std::env::args()
        .nth(1)
        .and_then(|s| s.parse().ok())
        .unwrap_or_else(std::process::id);

    println!("Instance ID: {}", instance_id);
    println!("Domain ID: 0 (default)\n");

    // Create participant with default discovery settings
    // This automatically enables SPDP multicast discovery
    println!("Creating DomainParticipant...");
    let participant = hdds::Participant::builder(&format!("SimpleDiscovery_{}", instance_id))
        .domain_id(0)
        .with_transport(hdds::TransportMode::UdpMulticast)
        .build()?;

    println!("[OK] Participant created: {}", participant.name());
    println!("     GUID: {:?}", participant.guid());

    // Create a raw writer and reader for demonstration
    let writer = participant.create_raw_writer("DiscoveryDemo", None)?;
    println!("[OK] DataWriter created");

    let reader = participant.create_raw_reader("DiscoveryDemo", None)?;
    println!("[OK] DataReader created");

    println!("\n--- Discovery in Progress ---");
    println!("Waiting for other participants to join...");
    println!("(Run another instance of this sample to see discovery)\n");

    // Announce ourselves periodically
    let announce_interval = Duration::from_secs(2);
    let mut announce_count = 0;

    loop {
        // Send an announcement
        announce_count += 1;
        let message = format!(
            "Hello from instance {} (message #{})",
            instance_id, announce_count
        );

        match writer.write_raw(message.as_bytes()) {
            Ok(_) => println!("[SENT] {}", message),
            Err(e) => println!("[ERROR] Failed to send: {}", e),
        }

        // Small delay to allow data to arrive
        thread::sleep(Duration::from_millis(100));

        // Check for messages from other participants
        match reader.try_take_raw() {
            Ok(samples) => {
                for sample in samples {
                    if let Ok(data) = String::from_utf8(sample.payload) {
                        println!("[RECV] {}", data);
                    }
                }
            }
            Err(e) => println!("[WARN] Read error: {}", e),
        }

        // Wait before next announcement
        thread::sleep(announce_interval);

        // Exit after a while for demo purposes
        if announce_count >= 10 {
            println!("\n--- Sample complete (10 announcements sent) ---");
            break;
        }
    }

    // Show discovery statistics
    println!("\n--- Discovery Statistics ---");
    if let Some(discovery) = participant.discovery() {
        let count = discovery.participant_count();
        println!("Discovered {} other participant(s)", count);
    } else {
        println!("Discovery not available in IntraProcess mode");
    }

    println!("\n=== Sample Complete ===");
    Ok(())
}
